server/auth/auth.go

```diff
@@ -0,0 +1,358 @@
+		return fmt.Errorf("user %x not found", user[:])
+	}
+	return checkSigS256(msg, sig, client.acct.PubKey)
+}
+
+// Sign signs the msgjson.Signables with the DEX private key.
+func (auth *AuthManager) Sign(signables ...msgjson.Signable) error {
+	for i, signable := range signables {
+		sigMsg, err := signable.Serialize()
+		if err != nil {
+			return fmt.Errorf("signature message for signable index %d: %v", i, err)
+		}
+		sig, err := auth.signer.Sign(sigMsg)
+		if err != nil {
+			return fmt.Errorf("signature error: %v", err)
+		}
+		signable.SetSig(sig.Serialize())
+	}
+	return nil
+}
+
+// Send sends the non-Request-type msgjson.Message to the client identified by
+// the specified account ID.
+func (auth *AuthManager) Send(user account.AccountID, msg *msgjson.Message) {
+	client := auth.user(user)
+	if client == nil {
+		log.Errorf("Send requested for unknown user %x", user[:])
+		return
+	}
+	err := client.conn.Send(msg)
+	if err != nil {
+		log.Debugf("error sending on link: %v", err)
+	}
+}
+
+// Request sends the Request-type msgjson.Message to the client identified by
+// the specified account ID.
+func (auth *AuthManager) Request(user account.AccountID, msg *msgjson.Message, f func(*msgjson.Message)) {
+	client := auth.user(user)
+	if client == nil {
+		log.Errorf("Send requested for unknown user %x", user[:])
+		return
+	}
+	client.logReq(msg.ID, &respHandler{
+		expiration: time.Now().Add(reqExpiration),
+		f:          f,
+	})
+	err := client.conn.Request(msg, auth.handleResponse)
+	if err != nil {
+		log.Debugf("error sending request: %v", err)
+	}
+}
+
+// Penalize signals that a user has broken a rule of community conduct, and that
+// their account should be penalized.
+func (auth *AuthManager) Penalize(user account.AccountID, rule account.Rule) {
+	client := auth.user(user)
+	if client == nil {
+		log.Errorf("no client to penalize")
+		return
+	}
+	auth.storage.CloseAccount(client.acct.ID, rule)
+	client.conn.Banish() // May not want to do this. Leaving it for now.
+	auth.removeClient(client)
+}
+
+// user gets the clientInfo for the specified account ID.
+func (auth *AuthManager) user(user account.AccountID) *clientInfo {
+	auth.connMtx.RLock()
+	defer auth.connMtx.RUnlock()
+	return auth.users[user]
+}
+
+// conn gets the clientInfo for the specified connection ID.
+func (auth *AuthManager) conn(conn comms.Link) *clientInfo {
+	auth.connMtx.RLock()
+	defer auth.connMtx.RUnlock()
+	return auth.conns[conn.ID()]
+}
+
+// addClient adds the client to the users and conns maps.
+func (auth *AuthManager) addClient(client *clientInfo) {
+	auth.connMtx.Lock()
+	defer auth.connMtx.Unlock()
+	auth.users[client.acct.ID] = client
+	auth.conns[client.conn.ID()] = client
+}
+
+// removeClient removes the client from the users and conns map.
+func (auth *AuthManager) removeClient(client *clientInfo) {
+	auth.connMtx.Lock()
+	defer auth.connMtx.Unlock()
+	delete(auth.users, client.acct.ID)
+	delete(auth.conns, client.conn.ID())
+}
```

Comment:
Right now, the auth manager has no way of knowing when a client has disconnected, which isn't really a problem unless we're concerned about cleaning up the map of `clientInfo` structs. If that is a concern, I could handle that in one of two ways. 

1. Register another function with the comms module to receive notifications on disconnect.
2. Export the `(*wsLink).off` method and perform periodic clean up cycles.

Reply:
Let's put a TODO on it unless you've got something worked out.

Reply:
Sounds good. We will need a solution, but it will be easier once we put things together roughly and see how the dependencies develop.